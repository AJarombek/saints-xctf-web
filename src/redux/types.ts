/**
 * Types for the Redux store.
 * @author Andrew Jarombek
 * @since 7/23/2020
 */

export type RootState = {
    auth: AuthState,
    logs: LogsState
}

export type LogsState = {
    isFetching: boolean;
    didInvalidate: boolean;
    lastUpdated: number;
    items: Logs;
    feeds: LogFeeds;
    newComments: NewComments;
}

export type AuthState = {
    auth: Auth;
    user: Users;
}

export interface Meta {
    isFetching?: boolean;
    lastUpdated?: number;
    didInvalidate?: boolean;
    serverError?: string;
}

export type Auth = {
    isFetching?: boolean;
    lastUpdated?: number;
    signedInUser?: string;
    status?: string;
}

export type Users = {
    [key: string]: UserMeta;
}

export interface UserMeta extends User, Meta {}

export interface User {
    activation_code?: string;
    class_year?: number;
    deleted?: string;
    description?: string;
    email?: string;
    favorite_event?: string;
    first?: string;
    last?: string;
    last_signin?: string;
    location?: string;
    member_since?: string;
    password?: string;
    salt?: string;
    subscribed?: string;
    username?: string;
    week_start?: string;
}

enum LogType { RUN = "run", BIKE = "bike", SWIM = "swim", OTHER = "other" }
enum Metric { MILES = "miles", KILOMETERS = "kilometers", METERS = "meters" }

export type Log = {
    log_id: number;
    username: string;
    first: string;
    last: string;
    name: string;
    date: string;
    type: LogType;
    distance?: number;
    metric?: Metric;
    miles?: number;
    time?: string;
    pace?: string;
    feel: number;
    location?: string;
    description?: string;
    timeCreated?: string;
    comments?: Comment[];
}

export type Logs = {
    [key: string]: Record<string, Log>;
}

export type LogFeeds = {
    [key: string]: {
        filterBy: string;
        bucket: string;
        pages: LogFeedPages
    };
}

export type LogFeedPages = {
    [key: string]: LogFeedPage
}

export type LogFeedPage = {
    isFetching: boolean;
    lastUpdated: number;
    items: Log[];
    serverError: string;
}

export type Comment = {
    comment_id: number;
    username: string;
    first: string;
    last: string;
    log_id: number;
    time: string;
    content: string;
}

export type NewComments = {
    [key: string]: {
        isFetching: boolean;
        lastUpdated: number;
        created?: boolean;
        serverError?: string;
    }
}
