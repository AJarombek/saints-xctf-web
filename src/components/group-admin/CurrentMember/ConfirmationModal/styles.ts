/**
 * JSS styles for the ConfirmationModal component.
 * @author Andrew Jarombek
 * @since 1/17/2020
 */

import Mixins from '../../../../styles/mixins';

export default {
  modal: {
    ...Mixins.modal(),
  },
  modalButtons: {
    ...Mixins.modalButtons(),

    '& > .aj-contained-button': {
      marginRight: 10,
    },
  },
  disabledButton: {
    ...Mixins.disabledButton(),
  },
  spinner: {
    ...Mixins.buttonSpinner(),
  },
};
