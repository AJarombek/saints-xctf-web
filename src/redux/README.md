### Overview

Configure Redux as the state management system for the application.

### Files

| Filename                 | Description                                                                |
|--------------------------|----------------------------------------------------------------------------|
| `modules`                | Redux modules containing Reducers, Actions, and Action Creators.           |
| `appState.js`            | Sample application state maintained by Redux.                              |
| `store.js`               | Configuration and creation of the Redux store.                             |
| `types.ts`               | Typescript types used throughout the application.                          |

### Resources

1) [React/Redux Example App #1](https://github.com/erikras/react-redux-universal-hot-example/tree/master/src/redux)
2) [React/Redux Example App #2](https://github.com/goopscoop/ga-react-tutorial/tree/6-reduxActionsAndReducers/src/redux)
3) [React/Redux Example App #3](https://github.com/gothinkster/react-redux-realworld-example-app/tree/master/src)
4) [Redux Ducks Spec](https://github.com/erikras/ducks-modular-redux)
5) [React Router/Redux Setup Steps](https://github.com/supasate/connected-react-router#usage)
6) [Redux Hot Reloading](https://redux.js.org/recipes/configuring-your-store/#hot-reloading)
7) [Redux Thunk](https://github.com/reduxjs/redux-thunk)
8) [Redux/React Official Real World Example](https://github.com/reduxjs/redux/tree/master/examples/real-world)
