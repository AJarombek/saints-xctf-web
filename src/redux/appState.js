/**
 * Sample Redux State for the entire application.
 * @author Andrew Jarombek
 * @since 5/3/2020
 */

const appState = {
  auth: {
    auth: {
      isFetching: false,
      lastUpdated: 1588530309,
      signedIn: true,
      status: "SUCCESS"
    },
    registration: {
      stage: 1,
      first: "Andy",
      last: "Jarombek",
      email: "andrew@jarombek.com",
      username: undefined,
      password: undefined,
      activationCode: undefined,
      teams: {
        saintsXCTFAlumni: {
          status: "accepted",
          groups: {
            alumni: {
              status: "accepted"
            }
          }
        }
      }
    },
    user: {
      andy: {
        isFetching: false,
        didInvalidate: false,
        lastUpdated: 1588530310,
        username: "andy",
        first: "Andrew",
        last: "Jarombek",
        password: "xxxxxx",
        description: "Test Description",
        memberSince: "2016-02-23 12:00:00",
        classYear: 2017,
        location: "New York, NY",
        favoriteEvent: "Cooldown Jog",
        activationCode: "ABCD1234",
        email: "andrew@jarombek.com",
        lastSignIn: "2020-05-03 12:00:00",
        subscribed: "Y",
        logs: []
      }
    },
  },
  profile: {
    tom: {
      isFetching: false,
      didInvalidate: false,
      lastUpdated: 1588530311
    },
    joe: {
      isFetching: false,
      didInvalidate: false,
      lastUpdated: 1588530312
    },
    ben: {
      isFetching: false,
      didInvalidate: false,
      lastUpdated: 1588530313
    }
  },
  group: {
    wmenstf: {
      isFetching: false,
      didInvalidate: false,
      lastUpdated: 1588530314,
      groupName: "wmenstf",
      groupTitle: "Women's Track & Field",
      description: "",
      weekStart: "monday",
      members: [],
      logs: []
    }
  },
  logs: {
    isFetching: false,
    didInvalidate: false,
    lastUpdated: 1588530315,
    items: {
      1: {
        log_id: 1,
        username: "andy",
        first: "Andrew",
        last: "Jarombek",
        name: "Reservoir Loop",
        location: "New York, NY",
        date: "2020-05-03",
        type: "run",
        distance: 6.37,
        metric: "miles",
        miles: 6.37,
        time: "00:45:37",
        pace: "00:07:10",
        feel: 6,
        description: "",
        timeCreated: "2020-05-03 12:00:00",
        comments: {
          1: {
            commentId: 1,
            username: "andy",
            first: "Andrew",
            last: "Jarombek",
            logId: 1,
            time: "2020-05-03 12:00:00",
            content: "test"
          }
        }
      }
    }
  }
};
