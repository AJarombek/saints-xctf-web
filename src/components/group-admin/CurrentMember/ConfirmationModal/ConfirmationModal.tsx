/**
 * Component for a modal that confirms whether administrators want to remove or demote a user from a group.
 * @author Andrew Jarombek
 * @since 1/17/2021
 */

import React, { useMemo } from 'react';
import { createUseStyles } from 'react-jss';
import styles from './styles';
import { AJButton, AJModal } from 'jarombek-react-components';
import LoadingSpinner from '../../../shared/LoadingSpinner';
import { GroupMeta, MemberDetails, RootState } from '../../../../redux/types';
import { useSelector } from 'react-redux';

interface Props {
  onClose: () => void;
  onRemove: () => void;
  onDemote: () => void;
  show: boolean;
  isConfirming: boolean;
  member: MemberDetails;
  groupId: number;
}

const useStyles = createUseStyles(styles);

const ConfirmationModal: React.FunctionComponent<Props> = ({
  onClose,
  show,
  isConfirming,
  onRemove,
  onDemote,
  member,
  groupId
}) => {
  const classes = useStyles();

  const groups = useSelector((state: RootState) => state.groups?.group);

  const group: GroupMeta = useMemo(() => {
    return groups ? groups[groupId] : null;
  }, [groups, groupId]);

  if (show) {
    return (
      <AJModal backdrop={true} onClickBackground={onClose}>
        <div className={classes.modal} data-cypress="confirmationModal">
          {member.user === 'admin' && (
            <p>
              Are you sure you want to demote{' '}
              <b>
                {member.first} {member.last}{' '}
              </b>
              from their administrator position in <b>{group.group_title}</b>?
            </p>
          )}
          {member.user !== 'admin' && (
            <p>
              Are you sure you want to remove{' '}
              <b>
                {member.first} {member.last}{' '}
              </b>
              from <b>{group.group_title}</b>?
            </p>
          )}
          <div className={classes.modalButtons}>
            <AJButton
              type="contained"
              onClick={(): void => (member.user === 'admin' ? onDemote() : onRemove())}
              className={isConfirming && classes.disabledButton}
              disabled={isConfirming}
            >
              {member.user === 'admin' && <p>{isConfirming ? 'DEMOTING' : 'DEMOTE'}</p>}
              {member.user !== 'admin' && <p>{isConfirming ? 'REMOVING' : 'REMOVE'}</p>}
              {isConfirming && <LoadingSpinner className={classes.spinner} />}
            </AJButton>
            <AJButton type="outlined" onClick={onClose}>
              <p>CANCEL</p>
            </AJButton>
          </div>
        </div>
      </AJModal>
    );
  } else {
    return null;
  }
};

export default ConfirmationModal;
