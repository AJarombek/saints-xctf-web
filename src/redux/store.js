/**
 * Build and configure the redux store, which holds the application state and handles
 * all state updates.
 * @author Andrew Jarombek
 * @since 5/1/2020
 */

import { createBrowserHistory } from 'history';
import { createStore, applyMiddleware } from 'redux';
import { routerMiddleware } from 'connected-react-router';
import createLogger from 'redux-logger';
import reducer from './modules/reducers';

const history = createBrowserHistory();

const configureStore = (data) => {
  const loggerMiddleware = createLogger();
  const middleware = [loggerMiddleware, routerMiddleware(history)];

  const createStoreWithMiddleware = applyMiddleware(...middleware)(createStore);
  return createStoreWithMiddleware(reducer, data);
};

export { history };
export default configureStore;
