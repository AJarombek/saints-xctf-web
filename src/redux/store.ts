/**
 * Build and configure the redux store, which holds the application state and handles
 * all state updates.
 * @author Andrew Jarombek
 * @since 5/1/2020
 */

import { createBrowserHistory } from 'history';
import { createStore, applyMiddleware, Store, Action } from 'redux';
import { routerMiddleware } from 'connected-react-router';
import { createLogger } from 'redux-logger';
import thunkMiddleware, { ThunkAction } from 'redux-thunk';
import reducer from './modules/reducers';
import { RootState } from './types';

export const history = createBrowserHistory();

export default function configureStore(): Store {
  const loggerMiddleware = createLogger();
  const middleware = [routerMiddleware(history), loggerMiddleware, thunkMiddleware];

  const createStoreWithMiddleware = applyMiddleware(...middleware)(createStore);
  const store = createStoreWithMiddleware(reducer(history));

  if (process.env.NODE_ENV !== 'production' && module.hot) {
    module.hot.accept('./modules/reducers', () => store.replaceReducer(reducer));
  }

  return store;
}

export type AppThunk<ReturnType = void, State = RootState> = ThunkAction<ReturnType, State, unknown, Action<string>>;
