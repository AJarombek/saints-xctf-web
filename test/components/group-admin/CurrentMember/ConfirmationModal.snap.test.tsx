/**
 * Snapshot test for the ConfirmationModal component.
 * @author Andrew Jarombek
 * @since 3/5/2021
 */

import React from 'react';
import renderer from 'react-test-renderer';
import { Store } from 'redux';
import { emptyStore } from '../../../test-utils/storeMocks';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import ConfirmationModal from '../../../../src/components/group-admin/CurrentMember/ConfirmationModal';
import { alumni } from '../../../test-utils/groupMocks';

const mockStore = configureStore([thunk]);

describe('ConfirmationModal Snapshot Tests', () => {
  let store: Store;

  beforeEach(() => {
    store = mockStore({
      ...emptyStore,
      groups: {
        group: {
          1: alumni
        },
        members: {},
        stats: {},
        leaderboards: {},
        team: {},
        uploadingGroupPicture: {},
        updating: {}
      }
    });
  });

  it('renders correctly', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <ConfirmationModal
            onClose={(): void => null}
            onRemove={(): void => null}
            onDemote={(): void => null}
            show={true}
            isConfirming={false}
            member={{
              username: 'andy',
              first: 'Andy',
              last: 'Jarombek',
              member_since: '2016-12-23',
              user: 'admin',
              status: 'pending'
            }}
            groupId={1}
          />
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
