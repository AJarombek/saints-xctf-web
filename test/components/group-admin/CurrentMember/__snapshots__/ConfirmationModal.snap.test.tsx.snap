// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfirmationModal Snapshot Tests renders correctly 1`] = `
<div
  className="ajModalBackdrop-2-2-2 ajModalColorBackdrop-2-2-3 aj-modal-backdrop"
  onClick={[Function]}
>
  <div
    className="ajModal-2-2-1"
    onClick={[Function]}
  >
    <div
      className="modal-0-1-1"
    >
      <p>
        Are you sure you want to demote
         
        <b>
          Andy
           
          Jarombek
           
        </b>
        from their administrator position in 
        <b>
          Alumni
        </b>
        ?
      </p>
      <div
        className="modalButtons-0-1-2"
      >
        <div
          className="ajBaseButton-2-2-7 ajBaseButtonEnabled-2-2-8 aj-contained-button ajContainedButton-2-2-4 ajContainedButtonEnabled-2-2-5"
          onClick={[Function]}
        >
          <button
            disabled={false}
            type="button"
          >
            <p>
              DEMOTE
            </p>
          </button>
        </div>
        <div
          className="ajBaseButton-2-2-7 ajBaseButtonEnabled-2-2-8 aj-outlined-button ajOutlinedButton-2-2-10 ajOutlinedButtonEnabled-2-2-11"
          onClick={[Function]}
        >
          <button
            disabled={false}
            type="button"
          >
            <p>
              CANCEL
            </p>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
